### ISSUE TITLE
pixelation of jpg when using composite to overlay a SVG file

### ISSUE THREAD

**Issue by wyatt121** [libvips/libvips#1038]
Hi John,

I'm using your libraries to overlay SVG files over JPGs.  I'm getting quite a bit of pixelation on the edges of the SVG graphic overlay result.  Even with 100% quality jpegs.   

See these examples:  (red and green backgrounds are the worst)
https://images.hometownapparel.com/images/A1-RED-D2-WT-LT2.jpg
https://images.hometownapparel.com/images/A1-RED2-D2-WT-LT2.jpg  (100% quality jpg)
https://images.hometownapparel.com/images/A2-IGN-D2-WT-LT2.jpg

Any tips on how to stop this from happening?  Is there a solution with jpg, or is this just how jpg works?  Should I switch to a png base t-shirt file?

Any help is much appreciated.

Thank You,

Wyatt

**Comment by jcupitt** [comment]
Hi again, yes, JPG is designed for natural scenes, like the shirt. It won't work well on hard edges like the overlay. 

The simplest solution would just be to use PNG. I guess you could get fancy and store the shirt as JPG, the overlay as PNG, and then composite in the user's browser on display, but that would need some trickery and wouldn't work in all browsers.

**Comment by wyatt121** [comment]
Hi John,

If I use a source PNG, overlay the SVG and output a PNG I get perfect results.  No issues.  It's beautiful.   But PNG files are large files...  Not so good when dealing with millions of files.

If I use a source PNG, overlay the SVG and output as JPG, unfortunately, I still get the pixelation.  I guess I don't understand why...  The overlay is done on the png, not a jpg, and the output is basically being converted to a jpg...  I would think this would work.  

What am I missing?  Tell me why I'm wrong.  :)

Wyatt

**Comment by jcupitt** [comment]
Hiya, it's just the way that JPG works. 

When the SVG is rendered, the edges will be perfectly sharp. JPG compression is designed for natural scenes where all edges are somewhat fuzzy, so it gets its great compression ratio by subtly distorting edges and removing things you'd never notice. Give your beautiful hard edges to JPG and they will all get messed up.

You can stick with JPG for your source shirt image, but as soon as you add an unnatural super-hard edge, you'll need to use PNG if you want to keep perfect quality.

I think I would just use Q=90 JPG. It's not as good looking as PNG, but it's much, much smaller and faster.

**Comment by wyatt121** [comment]
Hi John,

Where do I set Q=90 for jpegs in relation to writeToBuffer?

I might just go with 8-bit PNG files.  I tested an 8-bit png source file, but unfortunately, writeToBuffer by default wrote out a 24-bit PNG file.  Therefore, I got a PNG file 2x as big as I had hoped.  Similarly to the jpg quality setting, you mentioned above, how do I define an 8-bit png file to be written out?

My current code:

```
$str = $out->writeToBuffer(".png");

// Display image to browser

header("Content-Type: image/png");
 echo $str;
exit();
```

Wyatt

**Comment by jcupitt** [comment]
Try:

```php
$str = $out->writeToBuffer(".jpg", ["Q" => 90]);
```

You can also use:

```php
$str = $out->writeToBuffer(".jpg[Q=90]");
```

if it's more convenient. 

8-bit PNG files are palette-ised, meaning they only contain 256 different colours, with other tones generated by dithering. This makes them look rather noisy. They are also very slow and memory-hungry to generate, since they need the extra optimisation step to pick the best 256 colours and do the dither. 

libvips 8.7 supports 8-bit PNG, but it's not out yet, unfortunately.

**Comment by wyatt121** [comment]
Hi John,

Ok.  I tried Q=90 and Q=100.    Both are good.  Q=100 is almost identical
to the PNG output.   Very little pixelation.  Hard to notice even when
zoomed in at 300%.

Thank you for your help!  I really appreciate it.  I wouldn't be able to
finish this project with you.

Wyatt

**Comment by wyatt121** [comment]
Hi John,

Just an update on my testing.  I ended up using JPG as my output set at Q=95.  This created a very small amount of pixelation around the edges that I found an acceptable trade-off for a significant reduction in file size, versus using Q=100.  I can only see it on red/green shirts and only if I try very hard to see it.  

I also experimented with using a PNG 24-bit source image, compared to a JPG 100% quality source image.  I saw very little difference in the output.  If anything, the jpeg output was slightly better.  In addition, the jpeg output processing was about 10% faster.

Thanks Again,

Wyatt
